File: leetcode/math/3247_Number_of_Subsequences_with_Odd_Sum/codes/dp(ver2).c
#include <stdio.h>

#define MOD 1000000007

int subsequenceCount(int* nums, int numsSize) {
    long long even = 0, odd = 0;

    for (int i = 0; i < numsSize; i++) {
        if (nums[i] % 2 == 0) {
            even = (even * 2 + 1) % MOD;
            odd = (odd * 2) % MOD;
        } else {
            long long new_even = (even + odd) % MOD;
            long long new_odd = (even + odd + 1) % MOD;
            even = new_even;
            odd = new_odd;
        }
    }

    return (int) odd;
}

File: leetcode/math/3247_Number_of_Subsequences_with_Odd_Sum/codes/dp(ver2).py
from typing import List

class Solution:
    def subsequenceCount(self, nums: List[int]) -> int:
        MOD = 10**9 + 7
        even, odd = 0, 0

        for num in nums:
            if num % 2 == 0:
                even = (even * 2 + 1) % MOD
                odd = (odd * 2) % MOD
            else:
                new_even = (even + odd) % MOD
                new_odd = (even + odd + 1) % MOD
                even, odd = new_even, new_odd

        return odd
